(#1251) Add limit parameter to Split

victornoel opened the pull request:
This is for #1251

0crat commented:
Job [#1263](https://github.com/yegor256/cactoos/issues/1263) is now in scope, role is `REV`

<!-- https://www.0crat.com/footprint/C63314D6Z/8fac2335-e003-4044-85bf-ca746d6204e0, version: 0.50.38, hash: ${buildNumber} -->

0crat commented:
This pull request [#1263](https://github.com/yegor256/cactoos/issues/1263) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/C63314D6Z/8db7ba65-3e3b-4e6a-a873-ba0c7276f167); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

<!-- https://www.0crat.com/footprint/C63314D6Z/69c38e99-9dff-42d8-aef9-24b52e889e2c, version: 0.50.38, hash: ${buildNumber} -->

paulodamaso commented:
@victornoel ping

victornoel commented:
@paulodamaso @iakunin I haven't forgotten, sorry for the delay, I will get back to it before end of week

victornoel commented:
@iakunin @paulodamaso sorry for the delay, it's done

paulodamaso commented:
@victornoel Build is failing, please take a look

victornoel commented:
@paulodamaso yes, sorry about that, it's fixed :)

paulodamaso commented:
@iakunin please go on with the review

victornoel commented:
@iakunin ping

victornoel commented:
@iakunin ping ?

paulodamaso commented on src/main/java/org/cactoos/text/Split.java:
```diff
@@ -45,6 +45,11 @@
     private final UncheckedText regex;
 
+    /**
+     * The limit.
```
@victornoel We need more documentation on how this `limit` works: how is the expected behavior with different limit values (> 0, == 0 and <0)

paulodamaso commented on src/test/java/org/cactoos/text/SplitTest.java:
```diff
@@ -92,4 +97,57 @@ public void splitTextWithTextRegex() throws Exception {
             new TextHasString("Split")
         ).affirm();
     }
+
```
@victornoel PLease add test cases for negative and zero limit values

victornoel commented:
@paulodamaso thx, it is done

victornoel commented:
@paulodamaso sorry, there was still some errors, now it is fixed

paulodamaso commented on src/test/java/org/cactoos/text/SplitTest.java:
```diff
@@ -23,17 +23,28 @@
+@SuppressWarnings(
+    {
+        "PMD.TooManyMethods",
+        "PMD.AvoidDuplicateLiterals"
```
@victornoel Make these duplicated literals constants and remove this supression

victornoel replied:
@paulodamaso I find that kind of strange, we do want independent tests: using constants is going against that and also the data used for tests is going to be declared far away frmo the tests. If you don't want the suppression, I will simply use different literals to avoid it. WDYT?

paulodamaso commented on src/test/java/org/cactoos/text/SplitTest.java:
```diff
@@ -92,4 +103,99 @@ public void splitTextWithTextRegex() throws Exception {
+        new Assertion<>(
+            "Must split a text with correct item",
+            new ItemAt<>(1, new Split("bar:and:foo", ":", -1)),
+            new ScalarHasValue<>(new TextIs("and"))
```
@victornoel wouldn't be better if we checked the exact value for this? What would be the expected result for `limit = -1`?

victornoel replied:
@paulodamaso I agree, I duplicated existing tests and didn't think to much about it. I will rewrite all the tests

paulodamaso commented on src/test/java/org/cactoos/text/SplitTest.java:
```diff
@@ -92,4 +103,99 @@ public void splitTextWithTextRegex() throws Exception {
+    public void splitTextZeroLimitItemValue() throws Exception {
+        new Assertion<>(
+            "Must split a text with correct item",
+            new ItemAt<>(1, new Split("bar:and:foo", ":", 0)),
```
@victornoel PLease look at the above

paulodamaso commented on src/test/java/org/cactoos/text/SplitTest.java:
```diff
@@ -92,4 +103,99 @@ public void splitTextWithTextRegex() throws Exception {
+        new Assertion<>(
+            "Must split a text with limited length and text regex",
+            new LengthOf(
+                new Split(new TextOf("fo:and:bo"), new TextOf(":"), 2)
```
@victornoel It is better to check for the elements, not the size of the array

victornoel replied:
@paulodamaso I agree, I duplicated existing tests and didn't think to much about it. I will rewrite all the tests

victornoel commented:
@paulodamaso I've taken your feedback into account, I will take care of it soon. As for the documentation, I put it a the head of the class (I delegated to `String.split` documentation to be precise)

victornoel commented:
@paulodamaso I've fixed the PR:
- rewrote the tests to be more complete in what they tests
- used `IterableEnvelope` in `Split` in order to have delegation of all the methods of the interface (I needed equals in particular)
- added doc on the constructors
- removed duplicate strings

paulodamaso commented:
@rultor merge

rultor commented:
> @rultor merge

@paulodamaso OK, I'll try to merge now. You can check the progress of the merge [here](https://www.rultor.com/t/21368-587030211)

rultor commented:
> @rultor merge

@paulodamaso Done! FYI, the full log is [here](https://www.rultor.com/t/21368-587030211) (took me 14min)

0crat commented:
Job [#1263](https://github.com/yegor256/cactoos/issues/1263) is not in the agenda of @iakunin[/z](https://www.0crat.com/u/iakunin), can't retrieve data and time of add

<!-- https://www.0crat.com/footprint/C63314D6Z/825d6749-3e52-4b76-bc68-5582860a2c64, version: 0.51.6, hash: ${buildNumber} -->

0crat commented:
Job [#1263](https://github.com/yegor256/cactoos/issues/1263) is not in the agenda of @iakunin[/z](https://www.0crat.com/u/iakunin), can't inspect

<!-- https://www.0crat.com/footprint/C63314D6Z/b96b6823-56b9-4fa3-bb87-50acfba1264a, version: 0.51.6, hash: ${buildNumber} -->

0crat commented:
Code review was too long (58 days), architects (@paulodamaso) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

<!-- https://www.0crat.com/footprint/C63314D6Z/26046790-82e6-4174-8633-dbb8549daa9e, version: 0.51.6, hash: ${buildNumber} -->

0crat commented:
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @iakunin[/z](https://www.0crat.com/u/iakunin), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/C63314D6Z/6ffcb08c-c58e-4dab-8274-2440e4ecd31f, version: 0.51.6, hash: ${buildNumber} -->

sereshqua commented:
@iakunin please make sure you will find at least 3 issues during next CR

0crat commented:
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

<!-- https://www.0crat.com/footprint/C63314D6Z/760d9132-df16-453f-9626-4ffe411cc81f, version: 0.51.6, hash: ${buildNumber} -->

iakunin commented:
@paulodamaso @victornoel sorry, guys, I’ve completely forgotten about this review, cause it’s been excluded from my Zerocracy agenda.

iakunin commented:
> @iakunin please make sure you will find at least 3 issues during next CR

@sereshqua got it.

sereshqua commented:
@0crat quality acceptable

0crat commented:
Quality review completed: +4 point(s) just awarded to @sereshqua[/z](https://www.0crat.com/u/sereshqua)

<!-- https://www.0crat.com/footprint/C63314D6Z/5dcd0c90-0135-414e-99d4-19c044663725, version: 0.51.6, hash: ${buildNumber} -->

0crat commented:
Order was finished, quality is "acceptable": +15 point(s) just awarded to @iakunin[/z](https://www.0crat.com/u/iakunin)

<!-- https://www.0crat.com/footprint/C63314D6Z/598c13de-56f1-4f78-944f-d7d081a27204, version: 0.51.6, hash: ${buildNumber} -->
